# DS 6600: Data Engineering: Lab 2

Jackson Miskill (jcm4bsq)

I did not use Generative AI for any portion of this assignment. 

In [85]:
import numpy as np
import pandas as pd
import requests
import json
from bs4 import BeautifulSoup
import dotenv
import os
dotenv.load_dotenv()

True

# 1a)

In [12]:
cands = pd.read_csv("data/cands22.txt", quotechar="|", header=None)
cands.columns = ['Cycle', 'FECCandID', 'CID','FirstLastP',
                 'Party','DistIDRunFor','DistIDCurr',
                 'CurrCand','CycleCand','CRPICO','RecipCode','NoPacs']
print(cands.shape)
cands

(8928, 12)


Cycle  FECCandID        CID                 FirstLastP Party  \
0      2022  H0AK00105  N00039029            Thomas Lamb (3)     3   
1      2022  H0AL01055  N00044245             Jerry Carl (R)     R   
2      2022  H0AL01063  N00044288            Wes Lambert (R)     R   
3      2022  H0AL01097  N00044750         James Averhart (D)     D   
4      2022  H0AL02087  N00030768            Martha Roby (R)     R   
...     ...        ...        ...                        ...   ...   
8923   2022  S8WI00224  N00041929            Leah Vukmir (R)     R   
8924   2022  S8WV00119  N00041293  Paula Jean Swearengin (D)     D   
8925   2022  S8WV00127  N00035531           Evan Jenkins (R)     R   
8926   2022  S8WV00135  N00041474              Bo Copley (R)     R   
8927   2022  S8WV00143  N00012642       Patrick Morrisey (R)     R   

     DistIDRunFor DistIDCurr CurrCand CycleCand CRPICO RecipCode NoPacs  
0            AK01                                             3N         
1            AL01       AL01        Y         Y      I        RW         
2            AL01                                             RN         
3            AL01                                             DN         
4            AL02                                             RN         
...           ...        ...      ...       ...    ...       ...    ...  
8923         WIS1                                             RN         
8924         WVS2                                             DN         
8925         WVS1                                             RN         
8926         WVS1                                             RN         
8927         WVS1                                             RN         

[8928 rows x 12 columns]

# 1b)

In [13]:
cand = pd.read_csv("data/cands22.txt", on_bad_lines='skip')
cand

|2022|  |H0AK00105|  |N00039029|            |Thomas Lamb (3)|  |3|  \
0     |2022|  |H0AL01055|  |N00044245|             |Jerry Carl (R)|  |R|   
1     |2022|  |H0AL01063|  |N00044288|            |Wes Lambert (R)|  |R|   
2     |2022|  |H0AL01097|  |N00044750|         |James Averhart (D)|  |D|   
3     |2022|  |H0AL02087|  |N00030768|            |Martha Roby (R)|  |R|   
4     |2022|  |H0AL02137|  |N00045120|          |Will Dismukes (R)|  |R|   
...      ...          ...          ...                          ...  ...   
8909  |2022|  |S8WI00224|  |N00041929|            |Leah Vukmir (R)|  |R|   
8910  |2022|  |S8WV00119|  |N00041293|  |Paula Jean Swearengin (D)|  |D|   
8911  |2022|  |S8WV00127|  |N00035531|           |Evan Jenkins (R)|  |R|   
8912  |2022|  |S8WV00135|  |N00041474|              |Bo Copley (R)|  |R|   
8913  |2022|  |S8WV00143|  |N00012642|       |Patrick Morrisey (R)|  |R|   

      |AK01|  |    |  | | | |.1 | |.2  |3N| | |.3  
0     |AL01|  |AL01|  |Y|   |Y|   |I|  |RW|   | |  
1     |AL01|  |    |  | |   | |   | |  |RN|   | |  
2     |AL01|  |    |  | |   | |   | |  |DN|   | |  
3     |AL02|  |    |  | |   | |   | |  |RN|   | |  
4     |AL02|  |    |  | |   | |   | |  |RN|   | |  
...      ...     ...  ...   ...   ...   ...   ...  
8909  |WIS1|  |    |  | |   | |   | |  |RN|   | |  
8910  |WVS2|  |    |  | |   | |   | |  |DN|   | |  
8911  |WVS1|  |    |  | |   | |   | |  |RN|   | |  
8912  |WVS1|  |    |  | |   | |   | |  |RN|   | |  
8913  |WVS1|  |    |  | |   | |   | |  |RN|   | |  

[8914 rows x 12 columns]

In [15]:
# pd.read_csv("data/cands22.txt")

This is likely not an appropriate way to load the data. The first and main problem I see with this is that this read dropped some of the rows. The dataframe should be 8928 rows and this dataframe is 8914 rows. I am guessing that the `on_bad_lines='skip'` essentially just drops records with data configurations that it did not understand. The error that comes from trying to read in the data without any separate configuration is that `ParserError: Error tokenizing data. C error: Expected 12 fields in line 851, saw 13`. I think the reason for this is that somewhere (likely in a name), there is an extra comma (,) where that comma is being treated like the separator instead of a comma in that field. Therefore, the parser reads an extra field (column) instead of continuing over it.

# 2

In [16]:
def get_useragent():
        url = "https://httpbin.org/user-agent"
        r = requests.get(url)
        useragent = json.loads(r.text)['user-agent']
        return useragent

def make_headers(email='jcm4bsq@virginia.edu'):
    useragent = get_useragent()
    headers = {
        'User-Agent': useragent,
        'From': email
    }
    return headers

base = "https://api.nasa.gov/"
endpoint = "neo/rest/v1/feed"
key = os.getenv("NASA_API_KEY")
params = {"api_key": key, 
          "start_date": "2024-10-08",
          "end_date":"2024-10-15"}
headers = make_headers()
r = requests.get(base + endpoint, params=params, headers=headers)



In [17]:
step1 = json.loads(r.text)

final_df = pd.DataFrame()
dates = list(step1['near_earth_objects'].keys())

for d in dates:
    temp = pd.json_normalize(step1['near_earth_objects'][d])
    final_df = pd.concat([final_df, temp])


print(final_df.shape)
final_df

(192, 18)


id neo_reference_id               name  \
0    2534676          2534676  534676 (2014 VK2)   
1    3826858          3826858          (2018 QE)   
2    3843427          3843427         (2019 OV3)   
3   54105558         54105558         (2021 AG6)   
4   54135427         54135427         (2021 GP3)   
..       ...              ...                ...   
11  54487066         54487066         (2024 TU1)   
12  54487280         54487280         (2024 TR2)   
13  54487290         54487290         (2024 TA3)   
14  54488494         54488494         (2024 TG5)   
15  54489677         54489677         (2024 TW8)   

                                         nasa_jpl_url  absolute_magnitude_h  \
0   https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 18.65   
1   https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 27.70   
2   https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 24.26   
3   https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 26.88   
4   https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 24.11   
..                                                ...                   ...   
11  https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 26.07   
12  https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 25.56   
13  https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 26.72   
14  https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 27.61   
15  https://ssd.jpl.nasa.gov/tools/sbdb_lookup.htm...                 27.03   

    is_potentially_hazardous_asteroid  \
0                               False   
1                               False   
2                               False   
3                               False   
4                               False   
..                                ...   
11                              False   
12                              False   
13                              False   
14                              False   
15                              False   

                                  close_approach_data  is_sentry_object  \
0   [{'close_approach_date': '2024-10-09', 'close_...             False   
1   [{'close_approach_date': '2024-10-09', 'close_...              True   
2   [{'close_approach_date': '2024-10-09', 'close_...             False   
3   [{'close_approach_date': '2024-10-09', 'close_...             False   
4   [{'close_approach_date': '2024-10-09', 'close_...             False   
..                                                ...               ...   
11  [{'close_approach_date': '2024-10-11', 'close_...             False   
12  [{'close_approach_date': '2024-10-11', 'close_...             False   
13  [{'close_approach_date': '2024-10-11', 'close_...             False   
14  [{'close_approach_date': '2024-10-11', 'close_...              True   
15  [{'close_approach_date': '2024-10-11', 'close_...             False   

                                           links.self  \
0   http://api.nasa.gov/neo/rest/v1/neo/2534676?ap...   
1   http://api.nasa.gov/neo/rest/v1/neo/3826858?ap...   
2   http://api.nasa.gov/neo/rest/v1/neo/3843427?ap...   
3   http://api.nasa.gov/neo/rest/v1/neo/54105558?a...   
4   http://api.nasa.gov/neo/rest/v1/neo/54135427?a...   
..                                                ...   
11  http://api.nasa.gov/neo/rest/v1/neo/54487066?a...   
12  http://api.nasa.gov/neo/rest/v1/neo/54487280?a...   
13  http://api.nasa.gov/neo/rest/v1/neo/54487290?a...   
14  http://api.nasa.gov/neo/rest/v1/neo/54488494?a...   
15  http://api.nasa.gov/neo/rest/v1/neo/54489677?a...   

    estimated_diameter.kilometers.estimated_diameter_min  \
0                                            0.494943      
1                                            0.007666      
2                                            0.037373      
3                                            0.011183      
4                                            0.040046    

# 3)

In [18]:
key = os.getenv("GENIUS_API_KEY")
headers = make_headers()
headers['Authorization'] = "Bearer " + key
headers

{'User-Agent': 'python-requests/2.32.3',
 'From': 'jcm4bsq@virginia.edu',
 'Authorization': 'Bearer v28gXb1_VUmcdbDYg1euMtA23I5diJzn953rhP2mAPiyMlrsqBiA0N_SXaA3-XKf'}

In [19]:
base = "https://api.genius.com"
endpoint = "/search"
params = {"q": "Lizzo"}
r = requests.get(base + endpoint, params = params, headers = headers)
lizzo_id = json.loads(r.text)['response']['hits'][0]['result']['primary_artist']['id']

new_endpoint = f"/artists/{lizzo_id}/songs"
r = requests.get(base + new_endpoint, params={"sort":"popularity", "per_page": 20}, headers=headers)

lizzo_df = pd.json_normalize(json.loads(r.text)['response']['songs'])
lizzo_df


annotation_count        api_path               artist_names  \
0                  9  /songs/3241792                      Lizzo   
1                 15  /songs/4198226                      Lizzo   
2                  4  /songs/2438187                      Lizzo   
3                 12  /songs/7823996                      Lizzo   
4                 21  /songs/7066853            Lizzo & Cardi B   
5                 11  /songs/8616851                        SZA   
6                  3  /songs/3797161                      Lizzo   
7                  7  /songs/8080840                      Lizzo   
8                 11  /songs/4312315  Lizzo (Ft. Missy Elliott)   
9                  5  /songs/4286014                      Lizzo   
10                15  /songs/4312316                      Lizzo   
11                 1  /songs/4948719  Lizzo (Ft. Ariana Grande)   
12                 7  /songs/4464794                      Lizzo   
13                 4  /songs/9160872                      Lizzo   
14                10  /songs/3241766     Charli xcx (Ft. Lizzo)   
15                11  /songs/3721618                      Lizzo   
16                 5  /songs/7640097                      Lizzo   
17                 3  /songs/3201991                      Lizzo   
18                 5  /songs/4855379              KIDZ BOP Kids   
19                 5  /songs/8029110                      Lizzo   

                                           full_title  \
0                                Truth Hurts by Lizzo   
1                                      Juice by Lizzo   
2                               Good As Hell by Lizzo   
3                            About Damn Time by Lizzo   
4                           Rumors by Lizzo & Cardi B   
5                                          F2F by SZA   
6                                       Boys by Lizzo   
7                                     Grrrls by Lizzo   
8                  Tempo by Lizzo (Ft. Missy Elliott)   
9                             Cuz I Love You by Lizzo   
10                               Like a Girl by Lizzo   
11  Good as Hell (Remix) by Lizzo (Ft. Ariana Grande)   
12                                  Soulmate by Lizzo   
13                                      Pink by Lizzo   
14    Blame It on Your Love by Charli xcx (Ft. Lizzo)   
15                                    Jerome by Lizzo   
16                                   Special by Lizzo   
17                                  Water Me by Lizzo   
18                       Truth Hurts by KIDZ BOP Kids   
19                   2 Be Loved (Am I Ready) by Lizzo   

                           header_image_thumbnail_url  \
0   https://images.genius.com/713f196ed0f42f17fc2b...   
1   https://images.genius.com/2fbb007529b4962cb0d8...   
2   https://images.genius.com/9a2e5f20873abfc0b7ab...   
3   https://images.genius.com/d69c5f3c6f3b7505bb86...   
4   https://images.genius.com/07249b1846118211b81d...   
5   https://images.genius.com/cde31e780d76c90305bc...   
6   https://images.genius.com/91c735bc3cb0bfe91c4d...   
7   https://images.genius.com/733bcd99684324716989...   
8   https://images.genius.com/b735aaf3ce3992f757e4...   
9   https://images.genius.com/78997f81e1b165ac69db...   
10  https://images.genius.com/3eb5642187434d00e4f2...   
11  https://images.genius.com/ad70c15d6cdb24007903...   
12  https://images.genius.com/3eb5642187434d00e4f2...   
13  https://images.genius.com/ec297b8459ca95ffeef0...   
14  https://images.genius.com/6432cc927bb33993656b...   
15  https://images.genius.com/3eb5642187434d00e4f2...   
16  https://images.genius.com/733bcd99684324716989...   
17  https://images.genius.com/57d9d235fbe4192ade54...   
18  https://images.genius.com/1cb09ede2e20b2fd9425...   
19  https://images.genius.com/733bcd99684324716989...   

                                     header_image_url       id  \
0   https://images.genius.com/713f196ed0f42f17fc2b...  3241792   
1   https://images.genius.com/2fbb007529b4962cb0d8...  4198226   
2 

# 4a

The `robots.txt` file from `genius.com` does not have anything in it that says that we cannot scrape anything from the website. In fact, they have specific permissions in their file that says that Claude and ChatGPT can scrape their data. I am assuming this is because they want those bots to use the data on their website. It looks like they just don't want you to scrape data generated from their application (comments, metadata, email notifications, etc). 

In [63]:
def scrape_genius(url="https://genius.com/Lizzo-good-as-hell-lyrics"):
    r = requests.get(url)
    gumbo = BeautifulSoup(r.text, 'html.parser') 
    list_of_lyrics = gumbo.body.find_all("div", attrs={"data-lyrics-container": "true"})

    part1 = list_of_lyrics[0]
    part2 = list_of_lyrics[1]
    part3 = list_of_lyrics[2]

    lyrics = []
    for l in part1.contents:
        if str(l)[0] != "<":
            lyrics.append(l)

    for l in part2.contents:
        if str(l)[0] != "<":
            lyrics.append(l)

    for l in part3.contents:
        if str(l)[0] != "<":
            lyrics.append(l)

    return lyrics

In [64]:
other_tracks = gumbo.find_all("ol", attrs={"class": "AlbumTracklist__Container-sc-123giuo-0 kGJQLs"})

list_of_tracks = other_tracks[0].find_all("li")
print(len(list_of_tracks))
all_tracks = {}
for i in range(0, len(list_of_tracks)):
    try:
        print("Track", i)
        lyrics = scrape_genius(list_of_tracks[i].a['href'])
        all_tracks[i] = lyrics
    except:
        continue

print(all_tracks)




15
Track 0
Track 1
Track 2
Track 3
Track 4
Track 5
Track 6
Track 7
Track 8
Track 9
Track 10
Track 11
Track 12
Track 13
Track 14
{1: ['[Intro]', '[Verse 1]', 'Bad bitch, diamonds in my collarbone (Yee, yee)', 'Buy my whip by myself, pay my rent by myself', '[Pre-Chorus]', 'Got nothing to prove (Got nothing to prove)', "But I'ma show you how I do", '[Chorus]', 'Go and throw it (Like a girl)', "'Cause I run it (Like a girl)", 'Run it, run it (Like a girl)', 'I work my femininity', 'I make these boys get on their knees', 'Now watch me do it, watch me do it', "Look it, look it, I'ma do it", 'Like a girl (Like a girl)', 'Like a girl (Like a girl)', '[Verse 2]', 'Uh, put me on a pedestal, bet on me, bet I will', '[Pre-Chorus]', 'Got nothing to prove (Got nothing to prove)', "But I'ma show you how I do (But I'ma show you how I do)", '[Chorus]', "Find me up in Magic City bustin' hundreds by the bands", 'And I throw it (Like a girl)', 'Go and throw it (Like a girl)', "'Cause I run it (Like a gir

# 5

In [84]:
base = "https://maps.spirithalloween.com"
endpoint = "/api/getAsyncLocations"

params = {"template":"domain",
          "level":"domain"}

headers = make_headers()


spooky = requests.get(base+endpoint, params=params, headers=headers)
spooky_json = json.loads(spooky.text)


list_of_locations = []
for loc in spooky_json['markers']:
    gumbo = BeautifulSoup(loc['info'])
    tag = gumbo.find_all("div", attrs={"class": "address-two"})
    list_of_locations += tag[0].contents


final_list = []
for loc in list_of_locations:
    if "former" in loc.lower().strip():
        final_list.append(loc)

final_list
pd.Series(final_list).value_counts()

Former Sears                             3
Former Party City                        2
Former Rue 21                            2
Former CVS                               2
Former Party Depot                       2
Former Charming Charlies                 1
Former DSW Shoes                         1
Former Advance Auto                      1
Upper Floor of Former Lord and Taylor    1
Former Tuesday Morning                   1
Former Rite Aid                          1
Former Danker Furniture                  1
Former Joann                             1
Former Big Lots                          1
Former Annas Linens                      1
Former NAPA                              1
Former Sam Ash                           1
Former Dots                              1
Former Hair Salon                        1
Former Orvis Clothing                    1
Former Ace Hardware                      1
Former Kmart                             1
Name: count, dtype: int64